Comment Detail: Render content in web view

dvdchr opened the pull request:
Refs #17108, #17087, `aXjNk0P-zpl`

This adds the web view implementation to render the comment content. Some notes:
- Fixed a little bug when fetching the user's avatar URL.
- `richCommentTemplate.html` and `richCommentStyle.css` were added to stylize the contents according to the designs.
- Navigation requests from the web view are temporarily disabled for now. i.e. links look clickable but will do nothing (for now). It'll be implemented in the next PR.
- To handle content resize on orientation change, I ended up reloading the contents (i.e. calling `loadHTMLString` on the web view) again. 
  - I've tried other methods (`reload()`, through javascript's `location.reload()`, re-querying `scrollHeight`, ...) to no avail. The resize only works by calling `loadHTMLString` again. Somewhat expensive, I know. 😅 But it's what works so far and I've yet to find a better way.
- All resources within the comment content (images, video thumbnails) are loaded synchronously for now. For heavy comments (e.g. with >5 images), it may take a couple of seconds for the web view to fully render the content.
  - I'll take a look at loading the images asynchronously in the next PR. My assumption is it'll _at least_ cause the table to recalculate the cell height after each image completes loading, but if the method proves too tricky, I'll defer the async improvement and instead show a loading animation on top of the web view for the much simpler approach.
- Lastly, cell/component visibility based on user permission will be implemented in the upcoming PR.

## To test

- Ensure that the `New Comment Detail` flag is enabled.
- Go to My Site > Comments, and pick any comment.
- Verify that the comment content is displayed correctly (and beautifully ✨). 
  - Note that you may need to jump to multiple sites and hunt for tricky comments that may break the layout or appear strange. If you find them, please let me know! 
- Try changing to landscape orientation, and verify that the content is resized properly.

## Regression Notes
1. Potential unintended areas of impact
n/a. Feature is still hidden behind a flag.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
n/a. Feature is still hidden behind a flag.

3. What automated tests I added (or what prevented me from doing so)
n/a. Feature is still hidden behind a flag.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

dvdchr committed:
Update webview constraints

dvdchr committed:
Set dynamic font size on header labels

dvdchr committed:
Configure webView and load contents with richEmbedTemplate

dvdchr committed:
Calculate height and call completion block for caller to refresh height

dvdchr committed:
Fix avatar URL using wrong property

dvdchr committed:
minor: callback docs

dvdchr committed:
Cache the HTML template string

dvdchr committed:
Add html template for rich comment content

dvdchr committed:
Remove empty HTML tags from comment content

dvdchr committed:
Recalculate cell height on orientation change.

Also improve styling for <pre> and <blockquote>.

dvdchr committed:
Fix white flash for web view in dark mode

dvdchr committed:
Allow overflowing table to scroll horizontally

dvdchr committed:
Remove print statement

dvdchr committed:
Improve styling for image and code snippets

dvdchr committed:
Override hardcoded block styles

dvdchr committed:
Extract styles to a standalone CSS file.

dvdchr committed:
Fix resources location

dvdchr committed:
Patch blocks gallery grid appearance

dvdchr committed:
Reject all web view navigation requests for now.

dvdchr commented on WordPress/Classes/ViewRelated/Comments/CommentContentTableViewCell.swift:
```diff
@@ -32,36 +35,123 @@ class CommentContentTableViewCell: UITableViewCell, NibReusable {
+            return
+        }
+
+        webViewHeightConstraint.constant = 1
```
Forgot to add a code comment here, why I needed to reset the constraint before updating it with the new one. Will add this tomorrow.

dvdchr replied:
This line has been removed, so it's no longer needed.

dvdchr commented on WordPress/Classes/ViewRelated/Comments/CommentContentTableViewCell.swift:
```diff
@@ -168,4 +265,7 @@ private extension String {
                                                 + "%1$d is a placeholder for the number of Likes.")
+
+    // pattern that detects empty HTML elements (including HTML comments within).
+    static let emptyElementRegexPattern = "<[a-z]+>(<!-- [a-z\\/:]+ -->)+<\\/[a-z]+>"
```
I did find some parameters embedded in the HTML comments, e.g.
```html
<!-- wp:image {"id":10675,"sizeSlug":"large","linkDestination":"none"} -->
```
which escapes the regex pattern. Will fix this tomorrow!

dvdchr replied:
Addressed in https://github.com/wordpress-mobile/WordPress-iOS/pull/17129/commits/2c0ae96da04e1f7de9dedbce7851f4be27ca7ccc.

dvdchr committed:
Fix navigation policy to allow loading local files

dvdchr commented on WordPress/Classes/ViewRelated/Comments/CommentContentTableViewCell.swift:
```diff
@@ -32,36 +35,128 @@ class CommentContentTableViewCell: UITableViewCell, NibReusable {
+        }
+
+        webViewHeightConstraint.constant = 1
+        webView.loadHTMLString(cachedContent, baseURL: nil)
```
Should have passed `Self.resourceURL` here so the CSS file is properly loaded. Will fix this tomorrow!

dvdchr replied:
Addressed in https://github.com/wordpress-mobile/WordPress-iOS/pull/17129/commits/550a4d2732c9643ee2cd23e60bb09dafdcba6101 as part of the refactor to `viewWillTransition` from `traitCollectionDidChange`.

aerych commented on WordPress/Classes/ViewRelated/Comments/CommentContentTableViewCell.swift:
```diff
@@ -32,36 +35,128 @@ class CommentContentTableViewCell: UITableViewCell, NibReusable {
     }
+
+    override func traitCollectionDidChange(_ previousTraitCollection: UITraitCollection?) {
+        // when screen orientation changed, reload the HTML content to ensure that the content height accounts for the new width.
```
It looks like this won't be triggered in all situations.  I tried testing with an iPad Pro 9.7 simulator. The comment detail is shown in as the detail of a UISplitViewController.  In this scenario the size will change on rotation but the trait collection does not. 
Maybe this is something that could be done with `viewWillLayoutSubviews` or `viewDidLayoutSubviews`?

dvdchr replied:
Ah, good catch! In this case, I think I'll use `viewWillTransition` instead to detect size changes. 🙂 Will address this soon!

dvdchr replied:
Addressed in https://github.com/wordpress-mobile/WordPress-iOS/pull/17129/commits/550a4d2732c9643ee2cd23e60bb09dafdcba6101!

aerych commented on WordPress/Resources/HTML/richCommentStyle.css:
```diff
@@ -0,0 +1,189 @@
+
+/** IMAGES **/
+
+.wp-block-image img {
```
We might need to have a rule for regular `img` tags that aren't part of gutenberg blocks. I tested with a non-gutenberg comment that had an image tag. The portrait image had some issues with layout. The width was constrained to width of the web view but the aspect ratio was lost resulting in the image having it's natural height. 
I tried adding 
```
img {
    margin: 0 auto; /* align center */
    width: auto;
    max-width: 100vw;
    height: auto;
}
```
This seemed to fix the issue with the image tho I'm not sure if it's all necessary or would have other side effects we don't want.

dvdchr replied:
Good catch again! 🙂 Will fix this in the upcoming commit.

dvdchr replied:
Addressed in https://github.com/wordpress-mobile/WordPress-iOS/pull/17129/commits/1337847fbee3cd3240a33469a7a30c25b4d643f6.

aerych reviewed (changes requested):
Heya @dvdchr  👋 
Looking sharp!  I tested in in the iPhone Pro 11 simulator and the iPad Pro 9.7 simulator and things mostly looked good.  I did spot a couple of things and left comments about them. Lemme know what you think.  
Back to you sir!

aerych commented:
For some light stress testing I created [this comment](https://aerych.wpsandbox.me/2021/07/27/test-image/#comment-19) on a test self-hosted site that includes (I think) all of the default allowed tags for comments, plus images and YouTube embeds (which are not allowed by default).  The comment looks pretty hideous in a browser but it looks rather nice in the app. :) 
![Simulator Screen Shot - iPhone 11 Pro - 2021-09-07 at 18 11 51](https://user-images.githubusercontent.com/1435271/132421910-3162d904-0aa5-408a-a193-1da544f7ade1.png)
![Simulator Screen Shot - iPhone 11 Pro - 2021-09-07 at 18 11 55](https://user-images.githubusercontent.com/1435271/132421915-68913c31-3997-4dff-9dc8-fa9223ce8688.png)
![Simulator Screen Shot - iPhone 11 Pro - 2021-09-07 at 18 12 04](https://user-images.githubusercontent.com/1435271/132421917-f5760795-17f7-489a-84de-293933415630.png)


I did notice there is some trouble with the the YouTube embed sizing properly.  I'm not sure if that's because of the error screen or something else.  Something we'll want to look into.  Probably we'll have a period of finding lots of one-off quirky things (typical of this sort of change and to be expected) but we'll get 'em sorted and then we'll have some rock solid webview comments. :)

dvdchr committed:
Implement a smarter cache for the HTML string processing.

dvdchr committed:
Use viewWillTransition to react to orientation changes.

Additionally, the RowType enum is refactored; the .text case no longer provides a closure as associated value, in order to make the enum class conform to Equatable without manual implementation. Row actions are now implemented in the didSelectRow method instead.

dvdchr committed:
Disable shrinkToFit behavior on iPad

dvdchr committed:
Improve styling for several components.

* All elements now have max-width: 100vw to prevent content overflow, unless explicitly overridden (e.g. table). This should fix unexpected content overflows e.g. from YouTube embeds.
* Stylize img tags that aren't part of Gutenberg blocks.

dvdchr committed:
Set the webview back to opaque after content has finished loading.

The web view needs to be translucent initially, because otherwise the web view might show a short, white flash. This is noticeable in dark color schemes, and a tad annoying. On the other hand, a translucent webview causes scroll indicator to be invisible, regardless of the indicator style.

Therefore, the solution is to set the webview to translucent initially, and set it back to opaque once the content has finished loading.

dvdchr commented:
Appreciate the testing effort! 👍 Thank you.

> Probably we'll have a period of finding lots of one-off quirky things ...

Yeah... Also, considering how many blocks there are, I'm not sure if they would all look great, since the CSS is homegrown and thus, limited. 😆 

Additionally, I've found that embedded tweets are causing problems with height calculation since it asynchronously fetches Twitter's widget.js library and resizes the tweet component. This is rather tricky since I've yet to find a way to block async Javascript without affecting the Javascript code that's used to calculate document height. 

The content is still vertically scrollable, it still doesn't look good since the detail screen now has two vertically scrollable components. Although this only applies to comments with embedded Tweet blocks. I'll take a look at this in the next PR, and will address it if it's not taking up too much time. 

The rough idea is instead of using `evaluateJavascript`, I should use `WKUserContentController` and move the height calculation there. Then, I should be able to hook into Twitter's widget library and listen after the tweet component has finished rendering. Finally, call the height calculation function so it could trigger the table view to resize the cell. 

> I did notice there is some trouble with the YouTube embed sizing properly.

Yeah, for this I figured it's better to restrict everything from the start, and allow elements to scale past `100vw` explicitly (e.g. tables, custom emoji images):
```css
* {
  max-width: 100vw;
}
```
---

**Update**: this is now addressed in https://github.com/wordpress-mobile/WordPress-iOS/pull/17129/commits/1337847fbee3cd3240a33469a7a30c25b4d643f6.

dvdchr committed:
Increase background color visibility on mentions in dark mode

dvdchr committed:
Improve regex pattern to remove empty elements

dvdchr committed:
Improve component margins

dvdchr commented:
@aerych this is ready for a second round of review! 🙂

dvdchr committed:
Mark onContentLoaded property as private.

aerych commented on WordPress/Classes/ViewRelated/Comments/CommentDetailViewController.swift:
```diff
@@ -68,6 +68,16 @@ class CommentDetailViewController: UITableViewController {
+        super.viewWillTransition(to: size, with: coordinator)
+
+        // when an orientation change is triggered, recalculate the content cell's height.
+        guard let contentRowIndex = rows.firstIndex(where: { $0 == .content }) else {
```
Nice.

aerych reviewed (approved):
Heya @dvdchr 👋 
Changes look good! Confirming portrait images and youtube embeds are sizing better. :) The styles are preserved when rotating now and sizes are updated. 

![Simulator Screen Shot - iPhone 11 Pro - 2021-09-08 at 13 41 17](https://user-images.githubusercontent.com/1435271/132567743-ef624a97-4f73-4513-85b9-76a2eeb6e47a.png)

I noticed if I got very aggressive with rotation I could get into a state where the content height was higher than the actual content. I think it would be pretty hard to do this with an actual device so not a blocker imho.
![Simulator Screen Shot - iPhone 11 Pro - 2021-09-08 at 13 43 05](https://user-images.githubusercontent.com/1435271/132568285-26fba39b-82dd-4b48-b1ef-fd5f29c55187.png)

One thing we might add to a list of things to improve as we go is centering embedded content. What do you think? 
![Simulator Screen Shot - iPhone 11 Pro - 2021-09-08 at 13 41 56](https://user-images.githubusercontent.com/1435271/132568393-0cdb7f04-fd0b-44a7-8179-5ee755d729df.png)

Nice work sir.  Let's :shipit: !

dvdchr commented:
Thanks @aerych !

> I noticed if I got very aggressive with rotation I could get into a state where the content height was higher than the actual content. I think it would be pretty hard to do this with an actual device so not a blocker imho.

Good catch, and I agree with you. While this is not something crucial to address, I'll note it down with low priority for future reference. Or at least to keep the issue tracked.

> One thing we might add to a list of things to improve as we go is centering embedded content.

It looks like this is because the reply content is not using blocks. Here's an example HTML content:
```html
Here\'s an example of a YouTube embed:

<iframe width=\"560\" height=\"315\" src=\"https://www.youtube.com/embed/dQw4w9WgXcQ\" title=\"YouTube video player\" frameborder=\"0\" allow=\"accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture\" allowfullscreen></iframe>

Let\'s see how this will be rendered on the mobile side.
```

Unlike blocks, the texts are not wrapped in a paragraph element, thus kinda mucking up with the spacing. I'll try to explore if I can improve the spacing for plain text / non-Gutenberg block comments in the next PR.
